PIPELINE/Modify the docker file to fit with tensorflow gpu

ansonwong9695 opened the pull request.

ansonwong9695 committed:
Modify the docker file to fit with tensorflow gpu

nielsboecker commented on HoloPipelines/models/dense_vnet_abdominal_ct/Dockerfile:
```diff
@@ -0,0 +1,44 @@
+# Run install tensorflow, niftynet, flask
+RUN pip3 install --upgrade pip setuptools
+RUN pip3 install wheel
+RUN pip3 install -r requirements.txt
```
I spotted this error in the stdout, not sure if it is actually a problem? Installation continued afterwards:
```shell
ERROR: niftynet 0.5.0 has requirement numpy<=1.14.5,>=1.13.3, but you'll have numpy 1.17.0 which is incompatible.
```

ansonwong9695 replied:
I think this ERROR is fine, niftynet will later reinstall the numpy to a version that fits the requirement

nielsboecker commented on HoloPipelines/models/dense_vnet_abdominal_ct/Dockerfile:
```diff
@@ -0,0 +1,44 @@
+RUN pip3 install -r requirements.txt
+
+# Install abdominal model
+RUN net_download dense_vnet_abdominal_ct_model_zoo
```
When I try to build the image on my computer (WSL Ubuntu in Windows 10), it crashes at this point with the stacktrace below. Is this the error you encountered as well @ansonwong9695 ?

<details>
<summary>Show stacktrace</summary>

```shell
Step 12/17 : RUN net_download dense_vnet_abdominal_ct_model_zoo
 ---> Running in 9a532f38e74b
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/pywrap_tensorflow.py", line 58, in <module>
    from tensorflow.python.pywrap_tensorflow_internal import *
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/pywrap_tensorflow_internal.py", line 28, in <module>
    _pywrap_tensorflow_internal = swig_import_helper()
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/pywrap_tensorflow_internal.py", line 24, in swig_import_helper
    _mod = imp.load_module('_pywrap_tensorflow_internal', fp, pathname, description)
  File "/usr/lib/python3.5/imp.py", line 242, in load_module
    return load_dynamic(name, filename, file)
  File "/usr/lib/python3.5/imp.py", line 342, in load_dynamic
    return _load(spec)
ImportError: libcuda.so.1: cannot open shared object file: No such file or directory

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/niftynet/__init__.py", line 15, in <module>
    import tensorflow as tf
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/__init__.py", line 24, in <module>
    from tensorflow.python import pywrap_tensorflow  # pylint: disable=unused-import
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/__init__.py", line 49, in <module>
    from tensorflow.python import pywrap_tensorflow
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/pywrap_tensorflow.py", line 74, in <module>
    raise ImportError(msg)
ImportError: Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/pywrap_tensorflow.py", line 58, in <module>
    from tensorflow.python.pywrap_tensorflow_internal import *
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/pywrap_tensorflow_internal.py", line 28, in <module>
    _pywrap_tensorflow_internal = swig_import_helper()
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/pywrap_tensorflow_internal.py", line 24, in swig_import_helper
    _mod = imp.load_module('_pywrap_tensorflow_internal', fp, pathname, description)
  File "/usr/lib/python3.5/imp.py", line 242, in load_module
    return load_dynamic(name, filename, file)
  File "/usr/lib/python3.5/imp.py", line 342, in load_dynamic
    return _load(spec)
ImportError: libcuda.so.1: cannot open shared object file: No such file or directory


Failed to load the native TensorFlow runtime.

See https://www.tensorflow.org/install/errors

for some common reasons and solutions.  Include the entire stack trace
above this error message when asking for help.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/net_download", line 6, in <module>
    from niftynet.utilities.download import main
  File "/usr/local/lib/python3.5/dist-packages/niftynet/__init__.py", line 17, in <module>
    raise ImportError('NiftyNet is based on TensorFlow, which'
ImportError: NiftyNet is based on TensorFlow, which does not seem to be installed on your system.

Please install TensorFlow (https://www.tensorflow.org/) to be able to use NiftyNet.
The command '/bin/sh -c net_download dense_vnet_abdominal_ct_model_zoo' returned a non-zero code: 1
```
</details>

ansonwong9695 replied:
I think that is because you have not installed the Nvidia docker https://github.com/NVIDIA/nvidia-docker, 
after you installed you might also need to modify a JSON file by following this directory
 ```/etc/docker/daemon.json```
the json should be modified in this way
```
{
"default-runtime": "nvidia",
  "runtimes": {
     "nvidia": {
         "path": "/usr/bin/nvidia-container-runtime",
         "runtimeArgs": []
     }
 }

}
```
after the modification, you might need to restart the docker by running this two command lines
```
1. systemctl stop docker
2. systemctl start docker
```
you can test your modification by running this command line
```
docker run  --rm nvidia/cuda nvidia-smi
````
but I am not sure, this can solve your problem or not

nielsboecker replied:
Thanks Kawai, I will try tomorrow! If it works I will also copy your instructions to the `README`, this is the stuff we should document.

nielsboecker committed:
PIPELINE: Revert to using CPU

The tensorflow-gpu version depends on nvidia-docker runtime and causes several problems locally and on the CI environment, and is challenging as the first call after container startup takes up to 5 minutes, whereas it only improves general processing time by around 1.5 sec.

nbckr commented:
I am still getting the error for GPU, both local and on CI environment. Also there are a few other complications with the GPU version. This is a real blocker and therefore I decided to revert back to the CPU version for now. The image builds successfully and runs the test. Furthermore I adjusted the build pipeline so that it also builds this model's image.
